hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/SyncFuture.java

```diff
@@ -65,102 +74,128 @@
    *         if this future is already 'done' when this method called.
    */
-  synchronized boolean done(final long txid, final Throwable t) {
-    if (isDone()) {
-      return false;
-    }
-    this.throwable = t;
-    if (txid < this.txid) {
-      // Something badly wrong.
-      if (throwable == null) {
-        this.throwable =
-            new IllegalStateException("done txid=" + txid + ", my txid=" + this.txid);
+  boolean done(final long txid, final Throwable t) {
+    doneLock.lock();
+    try {
+      if (isDone()) {
+        return false;
       }
+      this.throwable = t;
+      if (txid < this.txid) {
+        // Something badly wrong.
+        if (throwable == null) {
+          this.throwable =
+              new IllegalStateException("done txid=" + txid + ", my txid=" + this.txid);
+        }
+      }
+      // Mark done.
+      this.doneTxid = txid;
+      doneCondition.signalAll();
+      return true;
+    } finally {
+      doneLock.unlock();
     }
-    // Mark done.
-    this.doneTxid = txid;
-    // Wake up waiting threads.
-    notify();
-    return true;
-  }
-
-  boolean cancel(boolean mayInterruptIfRunning) {
-    throw new UnsupportedOperationException();
   }
 
-  synchronized long get(long timeoutNs) throws InterruptedException,
+  long get(long timeoutNs) throws InterruptedException,
       ExecutionException, TimeoutIOException {
-    final long done = System.nanoTime() + timeoutNs;
-    while (!isDone()) {
-      wait(1000);
-      if (System.nanoTime() >= done) {
-        throw new TimeoutIOException(
-            "Failed to get sync result after " + TimeUnit.NANOSECONDS.toMillis(timeoutNs)
-                + " ms for txid=" + this.txid + ", WAL system stuck?");
+    doneLock.lock();
+    try {
+      while (!isDone()) {
+        if (!doneCondition.await(timeoutNs, TimeUnit.NANOSECONDS)) {
+          throw new TimeoutIOException("Failed to get sync result after "
+              + TimeUnit.NANOSECONDS.toMillis(timeoutNs) + " ms for txid=" + this.txid
+              + ", WAL system stuck?");
+        }
       }
+      if (this.throwable != null) {
+        throw new ExecutionException(this.throwable);
+      }
+      return this.doneTxid;
+    } finally {
+      doneLock.unlock();
     }
-    if (this.throwable != null) {
-      throw new ExecutionException(this.throwable);
-    }
-    return this.doneTxid;
-  }
-
-  synchronized boolean isDone() {
-    return this.doneTxid != NOT_DONE;
   }
 
-  synchronized boolean isThrowable() {
-    return isDone() && getThrowable() != null;
+  boolean isDone() {
+    if (doneLock == null) {
+      // First reset
+      return true;
+    }
+    doneLock.lock();
+    try {
+      return this.doneTxid != NOT_DONE;
+    } finally {
+      doneLock.unlock();
+    }
   }
 
-  synchronized Throwable getThrowable() {
-    return this.throwable;
+  Throwable getThrowable() {
+    doneLock.lock();
+    try {
+      if (!isDone()) {
```

Comment:
Better to just test this.doneTxid != NOT_DONE? Or introduce an unlocked isDone version may be called isDoneWithoutLock to call it in both isDone and getThrowable, so we do not need to lock in the isDone again here.

Reply:
It is a ReentrantLock, so in the same thread context, it doesn't acquire again?

Reply:
Also, if you are saying we can access doneTxid without a lock for specific codepaths, long operations per JVM spec are not atomic since it is split into two 32bit operations, so we may see a corrupt value and hence can result in a wrong isDone() check?

Reply:
No, it will be acuired again, as you still need to call unlock one more time you call lock one more time. Reentrant means you can lock the lock multiple times in the same thread, but at least it will waste a bit more cycles.

I mean change the code like this:

```
boolean isDoneWithoutLock() {
  return this.doneTxid != NOT_DONE;
}

boolean isDone() {
  doneLock.lock();
  try {
    return inDoneWithoutLock();
  } finally {
    doneLock.unlock();
  }
}

boolean done() {
  doneLock.lock();
  try {
    if (inDoneWithoutLock()) {
      return false;
    }
    ...
  } finally {
    doneLock.unlock();
  }
}

long get() {
  doneLock.lock();
  try {
    while (!inDoneWithoutLock()) {
      ...
    }
    ...
  } finally {
    doneLock.unlock();
  }
}
```

Reply:
We do not include modifiers here, just an example, not the final code. The isDoneWithoutLock should be private...

Reply:
Ok, re-entrant acquisition of the lock should be very low penalty but yes doing this way shaves off a few cycles, will do.